llms/openai/openai.go

```diff
@@ -0,0 +1,294 @@
+	if err != nil {
+		return "", err
+	}
+
+	return generations.Generations[0][0].Text, nil
+}
+
+func (openai *OpenAI) Generate(ctx context.Context, prompts []string, stop []string) (*llms.LLMResult, error) {
+	subPrompts := llms.BatchSlice[string](prompts, openai.batchSize)
+	maxTokens := openai.maxTokens
+	var completionTokens, promptTokens, totalTokens int64
+	var choices []shared.CreateCompletionResponseChoices
+
+	if openai.maxTokens == -1 {
+		if len(prompts) != 1 {
+			return nil, errors.New("max_tokens set to -1 not supported for multiple inputs")
+		}
+
+		maxTokens = llms.CalculateMaxTokens(prompts[0], openai.modelName)
+	}
+
+	if len(stop) == 0 {
+		stop = openai.stop
+	}
+
+	for _, prompts := range subPrompts {
+		data, err := openai.completionWithRetry(ctx, prompts, maxTokens, stop)
+		if err != nil {
+			// TODO: Wrap Into Informative Errors
+			return nil, err
+		}
+
+		choices = append(choices, data.Choices...)
+		if data.Usage != nil {
+			completionTokens += data.Usage.CompletionTokens
+			promptTokens += data.Usage.PromptTokens
+			totalTokens += data.Usage.TotalTokens
+		}
+	}
+	var generations [][]llms.Generation
+	batchedChoices := llms.BatchSlice[shared.CreateCompletionResponseChoices](choices, openai.n)
+	for _, batch := range batchedChoices {
+		var generationBatch []llms.Generation
+		for _, choice := range batch {
+			generationBatch = append(generationBatch, llms.Generation{
+				Text: *choice.Text,
+				GenerationInfo: map[string]interface{}{
+					"finishReason": choice.FinishReason,
+					"logprobs":     choice.Logprobs,
+				},
+			})
+		}
+		generations = append(generations, generationBatch)
+	}
+
+	return &llms.LLMResult{
+		Generations: generations,
+		LLMOutput: map[string]interface{}{
+			"completionTokens": completionTokens,
+			"promptTokens":     promptTokens,
+			"totalTokens":      totalTokens,
+		},
+	}, nil
+}
+
+func (openai *OpenAI) completionWithRetry(ctx context.Context, prompts []string, maxTokens int64, stop []string) (*shared.CreateCompletionResponse, error) {
+	promptRequest := shared.CreateCreateCompletionRequestPromptArrayOfstr(prompts)
+	request := shared.CreateCompletionRequest{
+		Model:            openai.modelName,
+		Prompt:           &promptRequest,
+		MaxTokens:        &maxTokens,
+		Temperature:      &openai.temperature,
+		TopP:             &openai.topP,
+		N:                &openai.n,
+		BestOf:           &openai.bestOf,
+		LogitBias:        openai.logitBias,
+		PresencePenalty:  &openai.presencePenalty,
+		FrequencyPenalty: &openai.frequencyPenalty,
+	}
+	if len(stop) != 0 {
+		stopRequest := shared.CreateCreateCompletionRequestStopArrayOfstr(stop)
+		request.Stop = &stopRequest
+	}
+
+	var finalResult *shared.CreateCompletionResponse
+	var finalErr error
+
+	// wait 2^x second between each retry starting with
+	// max 10 seconds
+	for i := 0; i < openai.maxRetries; i++ {
+		lastTry := i == openai.maxRetries-1
+		res, err := openai.client.OpenAI.CreateCompletion(ctx, request)
```

Comment:
Looks like we can't access error responses from OpenAI here. The only thing that gets marshaled by the SDK is 200 responses.

Reply:
Do you think it makes sense to leave the body open when returning RawResponse, or potentially copy to a new reader. I think that's a change we could make pretty easily.